### PR TITLE
plexamp: add ozone platform hint for wayland

### PR DESCRIPTION
**Pull request by jackwilsdon**
## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/applications/audio/plexamp/default.nix [hunk 1]
```diff
@@ -1,4 +1,4 @@
-{ lib, fetchurl, appimageTools }:
+{ lib, fetchurl, appimageTools, makeWrapper }:
```
Comment (JohnRTitor) [inline pkgs/applications/audio/plexamp/default.nix:1]:
```suggestion
{ lib, fetchurl, appimageTools, makeBinaryWrapper, }:
```

File: pkgs/applications/audio/plexamp/default.nix [hunk 2]
```diff
@@ -23,6 +23,9 @@ appimageTools.wrapType2 {
       $out/share/icons/hicolor/scalable/apps/plexamp.svg
     substituteInPlace $out/share/applications/${pname}.desktop \
       --replace 'Exec=AppRun' 'Exec=${pname}'
+    source "${makeWrapper}/nix-support/setup-hook"
```
Comment (JohnRTitor) [inline pkgs/applications/audio/plexamp/default.nix:26]:
```suggestion
```

**Review by JohnRTitor — COMMENTED** [review]
makeBinaryWrapper is recommended and just add it to nativeBuildInputs

**Comment by jackwilsdon** [comment]
Plexamp uses `appimageTools.wrapType2` which doesn't seem to accept `nativeBuildInputs` (it forwards its args to `buildFHSEnv`) - I copied sourcing the setup hook from another app image because of this.

I can change it to source the binary wrapper setup hook, but it seems like it needs `set +u` before `wrapProgram` as otherwise I get:
```text
/nix/store/hpdx3zx4jxpk8qw2ggk2aqfa4y1npja0-make-binary-wrapper-hook/nix-support/setup-hook: line 152: flagsBefore: unbound variable
```

**Comment by jackwilsdon** [comment]
I've done some testing and it doesn't look like we can use `makeBinaryWrapper` due to #172583 (we need variable expansion at runtime to set the flag only for Wayland).

**Review by JohnRTitor — APPROVED** [review]
Will merge it when I get back.

**Comment by JohnRTitor** [comment]
Can't seem to run this on X? How are you testing this? I am testing this on Hyprland.

Here's what I did.

```bash
WAYLAND_DISPLAY= plexamp
```

Relevant logs:
```
[274703:1016/165349.425471:ERROR:wayland_connection.cc(193)] Failed to connect to Wayland display: No such file or directory (2)
[274703:1016/165349.425500:ERROR:ozone_platform_wayland.cc(259)] Failed to initialize Wayland platform
[274703:1016/165349.425506:ERROR:env.cc(257)] The platform failed to initialize.  Exiting.
```

**Comment by jackwilsdon** [comment]
I can't reproduce your issue under Sway:

```text
$ NIXPKGS_ALLOW_UNFREE=1 WAYLAND_DISPLAY= nix run --impure .#plexamp
```

Makes Plexamp open under Xwayland (confirmed with `nix run nixpkgs#xorg.xlsclients`). You can confirm what flags are being passed using `bash -x plexamp`:

```text
$ bash -x ./result/bin/plexamp 
+ exec -a ./result/bin/plexamp /nix/store/3s8h2vldw5mg587hjgmmhvkxh9r2cys3-plexamp-4.11.2/bin/.plexamp-wrapped --ozone-platform-hint=auto --enable-features=WaylandWindowDecorations
```

```text
$ WAYLAND_DISPLAY= bash -x ./result/bin/plexamp 
+ exec -a ./result/bin/plexamp /nix/store/3s8h2vldw5mg587hjgmmhvkxh9r2cys3-plexamp-4.11.2/bin/.plexamp-wrapped
```

**Comment by JohnRTitor** [comment]
Looks like a Hyprland problem then, running on XWayland somehow is buggy and does not work sometimes.

Same thing happened with https://github.com/NixOS/nixpkgs/pull/346189#issuecomment-2391797898. So CC @fufexan, could you take a look?

```
❯  hyprctl version -j
{
    "branch": "",
    "commit": "6ae89940c761d4ed4317df0af29e8df41d472091",
    "dirty": false,
    "commit_message": "",
    "commit_date": "2024-10-09",
    "tag": "",
    "commits": "5323",
    "buildAquamarine": "0.4.2",
    "flags": []
}

```

**Comment by JohnRTitor** [comment]
As for this PR, let's merge this.

**Comment by fufexan** [comment]
> Looks like a Hyprland problem then, running on XWayland somehow is buggy and does not work sometimes.
> 
> Same thing happened with [#346189 (comment)](https://github.com/NixOS/nixpkgs/pull/346189#issuecomment-2391797898). So CC @fufexan, could you take a look?
> 
> ```
> ❯  hyprctl version -j
> {
>     "branch": "",
>     "commit": "6ae89940c761d4ed4317df0af29e8df41d472091",
>     "dirty": false,
>     "commit_message": "",
>     "commit_date": "2024-10-09",
>     "tag": "",
>     "commits": "5323",
>     "buildAquamarine": "0.4.2",
>     "flags": []
> }
> ```

Just checked, seems to work for me.

As for the linked issue, I couldn't get it to work
```
› DISPLAY=:0 NIXOS_OZONE_WL= WAYLAND_DISPLAY= wire-desktop 
[613079:1016/225518.938028:ERROR:wayland_connection.cc(196)] Failed to connect to Wayland display: Connection refused (111)
[613079:1016/225518.938077:ERROR:ozone_platform_wayland.cc(264)] Failed to initialize Wayland platform
[613079:1016/225518.938083:ERROR:env.cc(257)] The platform failed to initialize.  Exiting.
zsh: segmentation fault (core dumped)  WAYLAND_DISPLAY= wire-desktop
```

**Pull request merged**